lib/ddtrace/logger.rb

```diff
@@ -10,30 +11,89 @@ class Logger < ::Logger
     def initialize(*args, &block)
       super
       self.progname = LOG_PREFIX
+
+      # Rate limit buckets
+      @buckets = {}
     end
 
     def add(severity, message = nil, progname = nil, &block)
       where = ''
+      c = nil
+      skipped_messages = ''
+
+      # Logging rate of 0 means log everything, so skip bucket checking
+      if Datadog.configuration.logging_rate > 0
+        # Get the call stack to determine who is logging
+        c = caller
+
+        # Get the rate limit bucket key
+        key = rate_limit_key(severity, progname, c)
+
+        # Get the currently existing bucket, if one exists
+        last_bucket = @buckets[key]
+
+        # Get the current time bucket
+        # DEV: This is used to determine which logging_rate period this log occurred in
+        #   e.g. (1557944138 / 60) = 25965735
+        current_time_bucket = Time.now.to_i / Datadog.configuration.logging_rate
+
+        # If no previous bucket exists or the time bucket has changed, then we can log
+        if last_bucket.nil? || last_bucket[:time_bucket] != current_time_bucket
+          # Get the previous number of skipped messages to append to message
+          unless last_bucket.nil? || last_bucket[:skipped].zero?
+            skipped_messages = ", #{last_bucket[:skipped]} additional messages skipped"
+          end
+
+          # We are in a new time bucket, and are not rate limited
+          # Reset the bucket for this key
+          @buckets[key] = { time_bucket: current_time_bucket,
+                            skipped: 0 }
+
+        # Otherwise, we are in an existing bucket we already logged for
+        else
+          # Increment the skipped count and return without logging anything
+          @buckets[key][:skipped] += 1
+          return
+        end
+      end
 
       # We are in debug mode, or this is an error, add stack trace to help debugging
       if debug? || severity >= ::Logger::ERROR
-        c = caller
+        c = caller if c.nil?
         where = "(#{c[1]}) " if c.length > 1
       end
 
       if message.nil?
         if block_given?
-          super(severity, message, progname) do
+          super(severity, "#{message}#{skipped_messages}", progname) do
```

Comment:
I think if message is nil it should still be nil. Instead its set as a empty string.
I think this is the source of the change in format.

Also this is really weird how its formatted here

Reply:
https://apidock.com/ruby/v2_5_5/Logger/add
> progname | Program name string. Can be omitted. Treated as a message if no message and block are given. -- | --

Worst api ever 😬

Reply:
oh wow, ok.... great find!